EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,277 @@
+
+<img src="https://raw.githubusercontent.com/0xProject/EIPs/master/EIPS/eip-eth_signTypedData/eth_signTypedData.png" width="500px">
+
+This EIP is a continuation of a discussion [here](https://github.com/ethereum/EIPs/pull/683).
+I offered this solution in [a comment](https://github.com/ethereum/EIPs/pull/683#issuecomment-327945854).
+It proposes a more general solution then the one offered [here](https://github.com/ethereum/EIPs/pull/693).
+
+## Motivation
+
+There is a whole range of higher level protocols emerging on Ethereum. State channels, 0x protocol, Login protocols, etc... For scalability reasons these protocols want to sign some protocol-specific data off-chain. They then use those signed messages to trigger on-chain or other actions with important consequences.
+
+* Transferring ETH
+* Transferring tokens
+* Transferring ownership
+* etc...
+
+
+The current `eth_sign` implementation allows signing of arbitrary data without specifying it's type or structure. Some signers assume that the data is a UTF-8 string. Some signers show it as a hex encoded string. This leads to user confusion since it's impossible to verify the message you're signing unless it's a plaintext UTF-8 string.
+
+<img src="https://raw.githubusercontent.com/0xProject/EIPs/master/EIPS/eip-eth_signTypedData/eth_sign.png"  width="500px">
+
+Calling `eth_sign` on Metamask displays the string to sign. If the user is signing the result of hashing a complex structure involving multiple critical pieces of information, the only way to verify what they are signing is to re-hash the same structure in an independent script and make sure the hashes match.
+
+Calling `personal_sign` on Metamask with the raw bytes of a hash (e.g. not an ASCII string) shows the user this even less verifiable message that they should sign.
+
+<img src="https://raw.githubusercontent.com/0xProject/EIPs/master/EIPS/eip-eth_signTypedData/personal_sign.png"  width="500px">
+
+The main problem is that signers don't have enough metadata to display the DApp's real intent effectively. In order to do so, they need the plaintext input that the dApp wishes hashed and signed by the user.
+
+## Specification
+
+This EIP proposes a new JSON RPC method to the `eth` namespace: `eth_signTypedData`.
+
+Parameters:
+0. `TypedData` - Typed data to be signed
+1. `Address` - 20 Bytes - Address of the account that will sign the messages
+
+Returns:
+0. `DATA` - signature - 65-byte data in hexadecimal string
+
+Typed data is the array of data entries with their specified type and human-readable name. Below is the [json-schema](http://json-schema.org/) definition for `TypedData` param.
+```json-schema
+{
+  items: {
+    properties: {
+      name: {type: 'string'},
+      type: {type: 'string'}, // Solidity type as described here: https://github.com/ethereum/solidity/blob/93b1cc97022aa01e7daa9816bcc23108bbe008b5/libsolidity/ast/Types.cpp#L182
```

Comment:
Please reference the ABI instead, for example https://solidity.readthedocs.io/en/v0.4.21/abi-spec.html#types (yes, we should make it a proper EIP), the link into the solidity source code is extremely cryptic.